Fix aws-cni chaining for IPv6 EKS clusters

mKeRix opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

The newly introduced [IPv6 mode for EKS](https://aws.amazon.com/de/blogs/containers/amazon-eks-launches-ipv6-support/) will create clusters in which only IPv6 addresses will be assigned to pods and services. Up until now, Cilium in chaining mode with aws-cni enforced that at least 1 IPv4 address needed to be available. This is not valid anymore, so this change will replace the logic allowing it to work given that either an IPv4 or an IPv6 address is available.

In addition to that, the IPv6 setup of EKS introduces a new plugin to the aws-cni config, which will handle IPv4 traffic through a host-local NAT. To support this the conflist setup has been reworked to simply append to a copy of the existing aws-cni conflist. This way saves us additional complexity in the script - otherwise we would have to somehow determine whether the plugin should be enabled and we'd need to call the AWS metadata service to get the host IP. Either way we go, we will need to introduce a new package into the image (jq for the former, curl or wget for the latter). Would appreciate feedback on this.

**There is an important caveat.** With this implementation the policies will only work for IPv6. The new plugin in aws-cni creates a new interface named `v4if0` on each container ([reasoning](https://github.com/aws/amazon-vpc-cni-k8s/blob/192f8472bd9524f739ef2b9d65d91a7ce3b04d09/cmd/egress-v4-cni-plugin/cni.go#L87-L104)). This interface is not tracked by Cilium in any way, i.e. IPv4 traffic is unrestricted. If you have any idea of how we could get this veth interface under Cilium control as well please let me know.

_Note:_ Unfortunately my machine does not let me install the necessary tooling to run the tests - but I did test the change in an actual EKS cluster and confirmed them to be working. When applying a policy one has to keep in mind that the `fe80::/10` CIDR range needs to be allowed for egress and ingress as per #16285.

Fixes: #18495

```release-note
Add support for aws-cni chaining in IPv6 EKS clusters
```

mKeRix committed:
cni: Allow aws-cni endpoints with only IPv6 addresses

Clusters that were created with the new EKS IPv6 feature
will not assign any IPv4 addresses within the cluster.

Fixes #18495

Signed-off-by: Heiko Rothe <me@heikorothe.com>

mKeRix committed:
cni: Retrieve aws-cni configuration from existing file

With IPv6 another plugin was introduced - since this plugin needs a bit
of configuration the first approach is to simply tag onto the existing
conflist. If the original aws-cni config cannot be found we revert to
the old behavior of simply using a pre-made config. On IPv6 clusters
will lead to IPv4 NAT not being available, cluster functionality itself
will work just fine though.

Signed-off-by: Heiko Rothe <me@heikorothe.com>

aanm commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -174,33 +174,9 @@ EOF
 	;;
 
 "aws-cni")
-	cat > "${CNI_CONF_NAME}" <<EOF
```
Some concerns here. If the file `10-aws.conflist.cilium_bak` does not exist on the node the Cilium CNI config will not be created on the node. Previously it was always created by Cilium. Is it possible to have a backup plan in case `10-aws.conflist.cilium_bak` does not exist and we create the file regardless?

mKeRix replied:
Good point. I'll check if I can get the old behavior back and still give people the opportunity to customize the necessary settings. Might also be interesting for people that want to disable the IPv4 NAT when running Cilium.

mKeRix replied:
Hey @aanm, sorry for taking so long to address your review comment. I got side-tracked by a few other things. I just submitted a change that restores the old behavior in case that the file does not exist. This seemed to be the most straightforward solution. It should work for both IPv4 and IPv6 clusters, on IPv6 clusters pods just won't have host-local IPv4 NAT available with the backup strategy.

aanm commented on images/runtime/Dockerfile:
```diff
@@ -35,7 +35,8 @@ ENV FORCE_BUILD=1
 # Update ubuntu packages to the most recent versions
 RUN apt-get update && \
-    apt-get upgrade -y
+    apt-get upgrade -y && \
```
This changes requires a couple of more steps as they are explained [here](https://docs.cilium.io/en/v1.11/contributing/development/images/#update-cilium-builder-and-cilium-runtime-images). Please follow them so that we have all image digests up-to-date. Thanks!

mKeRix replied:
Apologies for missing these steps. I did the first part of it - unfortunately I can't see the members of team build, so I don't know who to ping.

pchaigno replied:
You can simply post a message in #development on Slack in such cases and someone will help :slightly_smiling_face:

mKeRix committed:
cni: Wait for CNI config in aws-cni chaining mode

This change will exit the install script with an error if the main CNI
config does not exist yet in the aws-cni chaining mode. Installation
will be retried later, after Kubernetes restarts the container. This is
required to achieve consistent behavior, as the aws-node pod may be
launched after Cilium.

Signed-off-by: Heiko Rothe <me@heikorothe.com>

pchaigno commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -174,10 +174,41 @@ EOF
+}
+EOF
+	  ;;
+  fi
```
I think commit `cni: Provide default for aws-cni chaining config` should be merged with the previous commit to avoid the number of touched code lines (a pain when git blaming).

pchaigno replied:
Same for commit `cni: Fix install script format`.

pchaigno reviewed (approved):
Thanks for the contribution! I only have two minor comments below.

Don't hesitate to ask for help on Slack if you need it to update the images.

tklauser reviewed (approved):
LGTM once @pchaigno's review comments https://github.com/cilium/cilium/pull/18522#discussion_r815449764 and https://github.com/cilium/cilium/pull/18522#discussion_r815449921 are addressed.

sayboras commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -174,7 +191,12 @@ EOF
 "aws-cni")
-	cat > "${CNI_CONF_NAME}" <<EOF
+  if [ -f "${CNI_CONF_DIR}/10-aws.conflist.cilium_bak" ]; then
+    jq --argjson 'ENABLE_DEBUG' "$ENABLE_DEBUG" --arg 'LOG_FILE' "$LOG_FILE" \
```
TIL about argjon vs arg

sayboras commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -87,16 +87,13 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
-			}
-
-			if len(addrs) < 1 {
-				return fmt.Errorf("no address configured inside container")
```
while the current behaviour didn't take care of ipv6, it actually follow idea of throwing error if no address configured inside container i.e. fail fast if possible.

We can follow the same idea here i.e. if both addrs and addrsv6 are empty, just throw same error. The changes in this PR is fine as I know that you have a check for vethIP and vethIPv6 below.

sayboras reviewed (approved):
LGTM :rocket:

pchaigno commented on .github/workflows/build-images-ci.yaml:
```diff
@@ -147,7 +147,7 @@ jobs:
             quay.io/${{ github.repository_owner }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-race
           build-args: |
-            BASE_IMAGE=quay.io/cilium/cilium-runtime:eaa047d20a1736c81bcd4cab603f6c47c5508b21@sha256:ec480391acffe3c30a05b6c3b583f28caa9d20f25f33a650416454b29dabf642
+            BASE_IMAGE=quay.io/cilium/cilium-runtime:0da282e47af1907e17a1aa8dba3d41de6879ef62
```
That seems incorrect:
```
$ docker pull quay.io/cilium/cilium-runtime:0da282e47af1907e17a1aa8dba3d41de6879ef62
Error response from daemon: manifest for quay.io/cilium/cilium-runtime:0da282e47af1907e17a1aa8dba3d41de6879ef62 not found: manifest unknown: manifest unknown
```

mKeRix replied:
I think I was too eager when updating the PR. I've removed the image digest updates and force-pushed just the changes, so that the image can be built by the pipeline. It looks like I have to run through the full digest update process again, as the images changed in the mean time. Apologies for the chaos!

mKeRix replied:
Looks like everything is working again. :)

mKeRix committed:
update cilium-runtime

Signed-off-by: Heiko Rothe <me@heikorothe.com>
